Shipping Labels: Add "activate service packages" form.

hafizrahman opened the pull request:
# What this PR does

This PR adds a functionality to allow users to activate a service package (also known as predefined package), during shipping label creation process.

Fixes #3992

# Testing Instructions
0. Make sure the test site has Shipping Labels creation enabled, and have an Order where Shipping Labels can be created.
1. Delete all activated packages in wp-admin: `wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings`.
2. In the app, find the Order from step 0 and create a Shipping Label, complete the "Ship from" and "Ship to" steps, then continue with the "Packaging details" step.
3. Tap the "Package selected" spinner, this will open the "Add new package" screen. 
4. Switch to the "Service Package" tab. This will show a loading screen, and then finally show a list of active packages that can be activated.
5. Select a package (remember the category and name), tap Done.
6. If the package is saved properly,  the screen will be navigated back to the "Package Details" screen. Make sure the name of the recently activated package is shown in the "Package selected" spinner.
7. Tap the "Package selected" spinner. Make sure this opens the "Selected Package" screen, and the recently added package is listed there.
8. Tap the "Add New Package" button again. Open the "Service Package" tab and make sure that the recently activated package is no longer listed there.
9. Pick another package there, and tap "Done". This will again navigate back to the "Package Details" screen. Make sure the package name is now shown in the "Package selected" spinner.

See video below for the entire flow:

https://user-images.githubusercontent.com/266376/126443151-cce449f6-3147-4495-922d-a3d2d497886a.mp4



# Known limitations
- No unit tests yet. This will be added in a separate PR.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hafizrahman committed:
Starter files for this feature. No functionality yet.

hafizrahman committed:
Add repo method to fetch selectable service packages.

hafizrahman committed:
Get and save selectable packages in vm.

hafizrahman committed:
Pass selectable packages list to adapter.

Plus some bits of checkstyle fixes.

hafizrahman committed:
Convert packages list into item list with headings for display purposes.

hafizrahman committed:
Add layout file for selectable package.

The design might not be 100% accurate yet, still work-in-progress.

hafizrahman committed:
Display packages list on recycler view.

hafizrahman committed:
Handle radio button check event.

Also ensure that radio buttons are selected/deselected properly.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageAdapter.kt:
```diff
@@ -0,0 +1,106 @@
+
+    private sealed class ListItem {
+        data class Header(val title: String) : ListItem()
+        data class Package(val data: ShippingPackage, val isChecked: Boolean = false) : ListItem()
```
I would suggest that you change the `data` type to `ServicePackageUiModel` instead, and remove the `isChecked` property, as it's just a duplicate.
This way, you will be able to simplify the logic of `updateData`.

hafizrahman replied:
Thanks, that's a good idea.

hafizrahman committed:
Refactor data updating in adapter.

hafizrahman committed:
Refactor data updating in adapter once more.

hafizrahman committed:
Refactor adapter to be ListAdapter instead, to make use of DiffUtil.

hafizrahman committed:
Add Done button and handle package activation.

This already covers both success and failure cases.

hafizrahman committed:
Add caching for selectable service packages.

This reduces API calls, as well as to make sure the list no longer offers service packages that were recently activated.

hafizrahman committed:
Fix duplicate menu item bug.

hafizrahman committed:
Add Skeleton view to show during package loading.

hafizrahman commented:
This PR still has one case to cover (what to show when all packages have been selected and there's no selectable packages anymore). Since Peril bot has mentioned that this PR is already large, I'm splitting that to a separate issue #4458.

hafizrahman committed:
Add dimension units to service package item information.

hafizrahman committed:
Merge branch 'develop' into add/activate-service-package

hafizrahman committed:
Add loading dialog while API call is running.

hafizrahman committed:
Add info text at the top of service package form.

Also adds a bit of design changes to match Figma, as well as make skeleton view look better with the addition of info text.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelServicePackageAdapter.kt:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    fun updateData(uiModels: List<ServicePackageUiModel>) {
+        val items = uiModels
+            .groupBy { it.data.category }
+            .flatMap { entry ->
+                val list = mutableListOf<ListItem>()
+                list.add(ListItem.Header(entry.key))
+                list.addAll(entry.value.map { ListItem.Package(it) })
+                list
```
For displaying purposes, this converts the list of packages into a list with category header(s) displayed above its packages.

hafizrahman committed:
Bug fix to make sure selected packages cache value is correct.

This is for an edge case where a user has 1 remaining package to activate, then activates it, then comes back to the activate package screen.

Without this fix, the activate package screen will incorrectly list all items again.

With this fix, the activate package screen will correctly show empty view.

anitaa1990 reviewed (approved):
This looks good @hafizrahman! Nice work 🥇 ! Let me know if this is good to merge and I can merge it 👍

hafizrahman commented:
@anitaa1990 Thank you! This is good to go :)
